Add support for ignoring client generated IDs

nikajukic opened the pull request:
Solves issue #48 

Added configurable client_id_prefix parameter which is taken into consideration when parsing parameters. 

All IDs starting with client_id_prefix will be ignored. 

This feature gives you the option to create new AND update existing nested records on parent update. 

```
{
  "type": "multitracks",
  "attributes": {
    "title": "Multitrack"
  },
  "relationships": {
    "tracks": {
       "data": [
        {
          "type": "tracks",
          "id": "123"                           // Existing ID for existing resources
        },
        {
          "type": "tracks",
          "id": "cid_new_track"                 // Client ID for new resources -> needs to match ID in included below
        }
       ]
    }
  },
  "included": [
    {
      "id": "123",                              // Existing ID for existing resources
      "type": "tracks",
      "attributes": {
        "name": "Piano"
       }
    },
    {
      "id": "cid_new_track",                    // Client ID for new resources -> needs to match ID in relationships below
      "type": "tracks",
      "attributes": {
        "name": "Drums"
       }
    }
  ]
}
```

```
params.from_jsonapi

{
  "multitrack" => {
    "title" => "Multitrack",
    "tracks_attributes" => {
      "0" =>  {
         "id" => "123",
         "name" => "Piano"
      },
      "1" =>  {                                 // No ID is present, so ActiveRecord#update correctly creates the new instance
         "name" => "Drums"
      }
    }
  }
}
```

nikajukic committed:
Ignore client generated IDs when parsing params

nikajukic committed:
Change default client id prefix

nikajukic committed:
Handle integer IDs when checking for client id prefix

nikajukic committed:
Add tests for updating and creating children on update action

nikajukic committed:
Add client prefix translator spec

nikajukic committed:
Move checking if ID starts with client_id_prefix to BaseHandler

nikajukic committed:
Remove building included object to BaseHandler

nikajukic committed:
Extract included object base to a method

nikajukic committed:
Fix rubocop issues

nikajukic committed:
Use patch with rails fix instead of patch method

nikajukic committed:
Add post with rails fix instead of post method

greena13 commented:
Thanks for all your efforts, @nikajukic. This looks like a great solution and would solve our needs discussed in the original issue.

Thinking of other users of the library, however, having a predefined prefix may be an unnecessarily breaking change. Possibly a default value of `nil` with a guard clause would allow a smoother upgrade path for library users, while allowing those who want to opt-in to the behaviour, the ability to do so.

The current implementation also doesn't provide a way to opt _out_ of this behaviour, either.

Something like:

```ruby
def client_generated_id?(related_id)
  return false unless JsonApi::Parameters.client_id_prefix

  related_id.to_s.starts_with?(JsonApi::Parameters.client_id_prefix)
end
```

Presuming a prefix may also be unnecessarily prescriptive, and perhaps it would be better to allow specifying a lambda:

```ruby
module JsonApi
  module Parameters
    @ensure_underscore_translation = false
    @ignore_id_filter = (id) -> { false }

    class << self
      attr_accessor :ensure_underscore_translation
      attr_accessor :ignore_id_filter
    end
  end
end

# ...
def build_included_object(included_object, related_id)
  included_object_base(included_object).tap do |body|
    body[:id] = related_id unless JsonApi::Parameters.ignore_id_filter(related_id)
  # ...
```

greena13 reviewed (changes requested):
Thanks for all your efforts, @nikajukic. This looks like a great solution and would solve our needs discussed in the original issue.

Thinking of other users of the library, however, having a predefined prefix may be an unnecessarily breaking change. Possibly a default value of `nil` with a guard clause would allow a smoother upgrade path for library users, while allowing those who want to opt-in to the behaviour, the ability to do so.

The current implementation also doesn't provide a way to opt _out_ of this behaviour, either.

Something like:

```ruby
def client_generated_id?(related_id)
  return false unless JsonApi::Parameters.client_id_prefix

  related_id.to_s.starts_with?(JsonApi::Parameters.client_id_prefix)
end
```

Presuming a prefix may also be unnecessarily prescriptive, and perhaps it would be better to allow specifying a lambda:

```ruby
module JsonApi
  module Parameters
    @ensure_underscore_translation = false
    @ignore_id_filter = (id) -> { false }

    class << self
      attr_accessor :ensure_underscore_translation
      attr_accessor :ignore_id_filter
    end
  end
end

# ...
def build_included_object(included_object, related_id)
  included_object_base(included_object).tap do |body|
    body[:id] = related_id unless JsonApi::Parameters.ignore_id_filter(related_id)
  # ...
```

choosen commented:
Thanks for your opinion @greena13 
I would add some additional information
Instead of `cid_` prefix in examples, we should use `lid_` as it is stated in draft of json api 1.1 (draft but still better than custom 😄 ) https://jsonapi.org/format/1.1/#document-resource-object-identification

I think it's fair point to disable it as default for backward compatibility.

And we thought about renaming prefix to `ignore_ids_with_prefix`, but @greena13 `@ignore_id_filter` is also good.
I am not sure if we should provide lambda, such elasticity is not needed probably. 
What do you think issue#48 participants❔  🎛️

greena13 commented:
I'm happy with any of the possibilities you've outlined.

The minimal set of changes appears to be changing from `cid_` to `lid_` (with it being part of the draft standard, I rescind any concerns about introducing "breaking changes" that could not be disabled, and would probably just suggest a major increment instead).

greena13 commented:
Actually, we have some legacy clients for which the breaking change of moving from `cid_` to `lid_` would present a problem.

Retaining the ability to specify an alternative prefix would remain very valuable for us.

nikajukic commented:
Thanks for your comments and suggestions 👍 

I think the best solution is the one @greena13 provided, with setting a default `client_id_prefix` to `nil` and checking if it's defined. 
I also like the `ignore_ids_with_prefix` config name instead of `client_id_prefix`.

Regarding the lambda, I'm not sure if it's needed here. I think having the prefix config option is good enough, but if you insist, we could add a lambda filter as well.

greena13 commented:
> Regarding the lambda, I'm not sure if it's needed here. I think having the prefix config option is good enough, but if you insist, we could add a lambda filter as well.

To help focus the remaining conversation, I'll rescind my original suggestion for a lambda (it won't currently serve our needs and if absolutely needed, consumers of the library can override the appropriate method - as one would expect when deviating from the intended usage of a library).

greena13 commented:
@nikajukic have there been any updates on this since last month?

choosen commented:
@greena13 actually we should use `lid` as local id key instead of id with the prefix according to draft.
it solves some problems but requires some additional changes for frontend.
How do you see that change ?

greena13 commented:
@choosen this is a very good point and a distinction that was somehow missed (did the proposal _change_, or was I just not attentive enough the first time?)

I would _imagine_ at first thought, using a additional id field to match resources on (but then discard it) would be a different approach/algorithm that should be done independently and regardless of this issue. 

But then we still have this separate issue of wanting to migrate legacy clients that use a value prefix to determine local only IDs. It doesn't perhaps feel like it should be part of the critical functionality of the gem, but is there some way of separating out the algorithm for including ids into a method or module that can be easily extended and overridden, so we could provide users with instructions on how to support this legacy workaround if needed?

nikajukic committed:
Make nil default client_id_prefix

nikajukic committed:
Rename client_id_prefix to ignore_ids_with_prefix

nikajukic committed:
Move setting ignore_ids_with_prefix to beginning of example

nikajukic commented:
@greena13 @choosen sorry for the late response and feature update.

I've applied the changes you wanted:
- renamed client_id_prefix to ignore_ids_with_prefix
- set default value for ignore_ids_with_prefix config to nil
- updated tests

Can you please check how does it look now?

choosen commented on lib/jsonapi_parameters/parameters.rb:
```diff
@@ -1,7 +1,7 @@
   module Parameters
     @ensure_underscore_translation = false
-    @client_id_prefix = 'cid_'
+    @client_id_prefix = nil
```
I think explicit define is easier to understand that 👍

choosen commented on lib/jsonapi_parameters/default_handlers/base_handler.rb:
```diff
@@ -36,6 +36,8 @@ def included_object_base(included_object)
           end
 
           def client_generated_id?(related_id)
+            return false unless JsonApi::Parameters.client_id_prefix.present?
```
simpy ruby version is better in this case (I don't like overusing activeSupport if nil is default)
```
            return false unless JsonApi::Parameters.client_id_prefix
```

choosen replied:
oh it's fixed already 👍

choosen commented:
@greena13  what do you think ?

greena13 commented:
Thanks @nikajukic and @choosen . This looks good to me as a solution to the problem as originally stated, and this will likely prove a valuable tool for consumers of the library looking to upgrade or move to something more standard.

However, long term, I expect to move to a `lid` (separate attribute) based solution. Although waiting for the specification to exit draft status, and/or all of the various serialisation/deserialisation libraries used in the particular stack I'm working with, to support it, will take some time yet.

So 👍 as a likely useful stop-gap in the interim, and possibly for other uses I haven't anticipated in the long term.

choosen commented on README.md:
````diff
@@ -72,6 +72,126 @@ If you provide any related resources in the `relationships` table, this gem will
+         "name" => "Drums"
+      }
+    }
+  }
+}
+```
+
+In case of updating existing nested resources and creating new ones in the same request, client needs to generate IDs for new resources and use existing ones for existing resources. Client IDs will be removed from params.
+
+
+```
+{
+  "type": "multitracks",
+  "attributes": {
+    "title": "Multitrack"
+  },
+  "relationships": {
+    "tracks": {
+       "data": [
+        {
+          "type": "tracks",
+          "id": "123"                           // Existing ID for existing resources
+        },
+        {
+          "type": "tracks",
+          "id": "cid_new_track"                 // Client ID for new resources -> needs to match ID in included below
+        }
+       ]
+    }
+  },
+  "included": [
+    {
+      "id": "123",                              // Existing ID for existing resources
+      "type": "tracks",
+      "attributes": {
+        "name": "Piano"
+       }
+    },
+    {
+      "id": "cid_new_track",                    // Client ID for new resources -> needs to match ID in relationships below
````
````suggestion
Library will add relations to record. Nested resources will receive attributes from _included_ section by `id` match.
By default id field is treated as Client-Generated ID, so it will identify record in database.

If you want to add related object attributes without providing Client-Generated ID then 
you can specify ignore_ids_with_prefix:

```ruby
JsonApi::Parameters.ignore_ids_with_prefix = 'local_id_' # library default is nil
```

If defined, all IDs starting with `local_id_` will be removed from parsed params.

In case of creating new nested resources, client will need to generate local IDs sent in `relationships` and `included` parts of request.

```js
{
"type": "multitracks",
"attributes": {
  "title": "Multitrack"
},
"relationships": {
  "tracks": {
     "data": [
      {
        "type": "tracks",
        "id": "local_id_new_track"                 // Local ID for new resources -> needs to match ID in included below
      }
     ]
  }
},
"included": [
  {
    "id": "local_id_new_track",                    // Local ID for new resources -> needs to match ID in relationships above
    "type": "tracks",
    "attributes": {
      "name": "Drums"
     }
  }
]
}
```

```js
params.from_jsonapi

{
"multitrack" => {
  "title" => "Multitrack",
  "tracks_attributes" => {
    "0" =>  {                // No ID is present, so ActiveRecord#create correctly creates
       "name" => "Drums"     // the new instance without Client-Generated ID
    }
  }
}
}
```

In case of updating existing nested resources and creating new ones in the same request, client needs to generate local IDs for new resources and use existing ones for existing resources. Local IDs will be removed from params.


```js
{
  "type": "multitracks",
  "attributes": {
    "title": "Multitrack"
  },
  "relationships": {
    "tracks": {
       "data": [
        {
          "type": "tracks",
          "id": "123"                     // Existing ID for existing resources
        },
        {
          "type": "tracks",
          "id": "local_id_new_track"      // Local ID for new resources -> needs to match ID in included below
        }
       ]
    }
  },
  "included": [
    {
      "id": "123",                        // Existing ID for existing resources
      "type": "tracks",
      "attributes": {
        "name": "Piano"
       }
    },
    {
      "id": "local_id_new_track",         // Local ID for new resources -> needs to match ID in relationships above
```
````

choosen replied:
@nikajukic  Can I fix those and merge it ?

nikajukic replied:
@choosen Sorry, I thought you already did it above, now I see it was only a suggestion 🙈 

Sure, go for it!

choosen reviewed (commented):
Some small change to readme only. 
I think we can leave spec as they are. Don't have to align them with readme example

choosen commented:
[JSON:API v1.1](https://jsonapi.org/format/) was finalized September 30, 2022! 🎉
CC: @greena13
